rust/lance-index/src/scalar/zoned.rs

```diff
@@ -0,0 +1,890 @@
+                let last_offset =
+                    RowAddress::new_from_u64(row_addr_col.value(batch_offset + take - 1))
+                        .row_offset() as u64;
+
+                if zone_start_offset.is_none() {
+                    zone_start_offset = Some(first_offset);
+                }
+                zone_end_offset = Some(last_offset);
+
+                current_zone_len += take;
+                batch_offset += take;
+
+                if current_zone_len == zone_size {
+                    Self::flush_zone(
+                        &mut self.processor,
+                        &mut zones,
+                        fragment_id,
+                        &mut current_zone_len,
+                        &mut zone_start_offset,
+                        &mut zone_end_offset,
+                    )?;
+                }
+            }
+        }
+
+        if current_zone_len > 0 {
+            if let Some(fragment_id) = current_fragment_id {
+                Self::flush_zone(
+                    &mut self.processor,
+                    &mut zones,
+                    fragment_id,
+                    &mut current_zone_len,
+                    &mut zone_start_offset,
+                    &mut zone_end_offset,
+                )?;
+            } else {
+                self.processor.reset()?;
+            }
+        }
+
+        Ok(zones)
+    }
+
+    fn flush_zone(
+        processor: &mut P,
+        zones: &mut Vec<P::ZoneStatistics>,
+        fragment_id: u64,
+        current_zone_len: &mut usize,
+        zone_start_offset: &mut Option<u64>,
+        zone_end_offset: &mut Option<u64>,
+    ) -> Result<()> {
+        if *current_zone_len == 0 {
+            processor.reset()?;
+            *zone_start_offset = None;
+            *zone_end_offset = None;
+            return Ok(());
+        }
+
+        let start = zone_start_offset.unwrap_or(0);
+        let inferred_end =
+            zone_end_offset.unwrap_or_else(|| start + (*current_zone_len as u64).saturating_sub(1));
+        if inferred_end < start {
+            return Err(Error::invalid_input(
+                "zone row offsets are out of order",
+                location!(),
+            ));
+        }
+        let bound = ZoneBound {
+            fragment_id,
+            start,
+            length: (inferred_end - start + 1) as usize,
+        };
+        let stats = processor.finish_zone(bound)?;
+        zones.push(stats);
+        *current_zone_len = 0;
+        *zone_start_offset = None;
+        *zone_end_offset = None;
+        processor.reset()?;
+        Ok(())
+    }
+
+    /// Count how many consecutive rows belong to the same fragment starting from `start`.
+    /// Returns the number of rows before encountering a different fragment or reaching the end.
+    fn count_run_length(
+        row_addrs: &UInt64Array,
+        start: usize,
+        total: usize,
+        fragment_id: u64,
+    ) -> usize {
+        let mut idx = start;
```

Comment:
This function is uncessary. It could be replaced with something like
```rust
        (start..total)
            .take_while(|&idx| (row_addrs.value(idx) >> 32) == fragment_id)
            .count()
```